pdr_backend/lake/partitioned_data_store.py

```diff
@@ -0,0 +1,114 @@
+import hashlib
+import os
+import polars as pl
+import duckdb
+from enforce_typing import enforce_types
+
+
+class PartitionedDataStore:
+    @enforce_types
+    def __init__(self, base_directory=str):
+        self.base_directory = base_directory
+        self.duckdb_conn = duckdb.connect(
+            database=":memory:"
+        )  # Keep a persistent connection
+
+    @enforce_types
+    def _generate_view_name(self, base_path=str) -> str:
+        """
+        Generate a unique view name for a given base path.
+        @arguments:
+            base_path - The base path to generate a view name for.
+        @returns:
+            str - A unique view name.
+        """
+
+        hash_object = hashlib.md5(base_path.encode())
+        return f"dataset_{hash_object.hexdigest()}"
+
+    @enforce_types
+    def append_data(self, df: pl.DataFrame, dataset_identifier: str):
+        """
+        Append data to an in-memory dataset.
+        @arguments:
+            df - The Polars DataFrame to append.
+            dataset_identifier - A unique identifier for the dataset.
+        """
+
+        view_name = self._generate_view_name(self.base_directory + dataset_identifier)
+        # Register the DataFrame as an Arrow table temporarily
+        # arrow_table = df.to_arrow()
+        self.duckdb_conn.register(view_name, df)
+
+    @enforce_types
+    def finalize_data(self, dataset_identifier: str):
+        """
+        Finalizes the data by writing the in-memory dataset to partitioned Parquet files.
+        @arguments:
+            dataset_identifier - A unique identifier for the dataset.
+        """
+
+        view_name = self._generate_view_name(self.base_directory + dataset_identifier)
+        query = f"SELECT * FROM {view_name}"
+        result_df = self.duckdb_conn.execute(query).df()
+
+        # Convert to Polars DataFrame for partitioned write
+        df = pl.DataFrame(result_df)
+
+        lazy_df = df.lazy()
+        # Determine partitions based on a timestamp column, assuming 'timestamp' column exists
+        lazy_df = lazy_df.with_columns(
+            [pl.from_epoch("timestamp", time_unit="s").alias("date")]
+        )
+        df = lazy_df.collect()
+
+        partitions = (
```

Comment:
Something to keep in mind (**do not address now**) is that not every table will be partitioned the same way. Some may be by yy-mm-dd, others could be by 0xaddress or timestamp_slot.